### DISCUSSION TITLE
Distributed force on hyperelastic object

### DISCUSSION THREAD

**Discussion by KaiWolven** [adtzlr/felupe#861] [Q&A] [answered]
Hi,
I've been trying for some time to apply deformations to hyperelastic objects, which is quite easy using Felupe. However, know that I plan on more complex stuff, I wanted to apply forces directly to the top surface of an object, specifically at Z.max(). I noticed the existence of the fem.SolidBodyForce function, but I'm not sure if it can be used in this case or if there are any other possible options.
Thank you.

**Comment by adtzlr** [comment]
Hi, 

`SolidBodyForce` is for things like gravity (*body force*). Do you want to apply a concentrated force on a surface? You could have a look on [Ex. 8](https://felupe.readthedocs.io/en/latest/examples/ex08_shear.html#sphx-glr-examples-ex08-shear-py) which uses a `MultiPointConstraint`.  Consider the following examples:

> [!NOTE]  
> It's important to remove the MPC - center point from `mesh.points_without_cells`. In this array, all **hanging** points are collected and ignored to avoid errors. But we would like to include the degrees-of-freedom of this center point in this example.


```python
import felupe as fem
import numpy as np

mesh = fem.Cube(n=6)
mesh.update(points=np.vstack([mesh.points, [2, 0.5, 0.5]]))
mesh.points_without_cells = np.array([], dtype=int)

region = fem.RegionHexahedron(mesh)
field = fem.FieldContainer([fem.Field(region, dim=3)])
solid = fem.SolidBody(umat=fem.NeoHooke(mu=1, bulk=2), field=field)

fixed = dict(fixed=fem.Boundary(field[0], fx=0))
points = np.arange(mesh.npoints)[mesh.x == 1.0]
mpc = fem.MultiPointConstraint(field, points=points, centerpoint=-1)
force = fem.PointLoad(field, points=[-1])

table = fem.math.linsteps([0, 1], num=5, axis=0, axes=3)
ramp = {force: 1 * table}
step = fem.Step(items=[solid, mpc, force], ramp=ramp, boundaries=fixed)

job = fem.Job(steps=[step]).evaluate()
plotter = solid.plot("Principal Values of Cauchy Stress", plotter=mpc.plot())
plotter.show()
```

![image](https://github.com/user-attachments/assets/eec13239-821f-4fe6-9aac-daea4dde4e26)

It is also possible to apply the MPC only in the direction of the force, i.e.

```python
import felupe as fem
import numpy as np

mesh = fem.Cube(n=6)
mesh.update(points=np.vstack([mesh.points, [2, 0.5, 0.5]]))
mesh.points_without_cells = np.array([], dtype=int)

region = fem.RegionHexahedron(mesh)
field = fem.FieldContainer([fem.Field(region, dim=3)])
solid = fem.SolidBody(umat=fem.NeoHooke(mu=1, bulk=2), field=field)

boundaries = dict(
    left=fem.Boundary(field[0], fx=0),
    right=fem.Boundary(field[0], fx=2, skip=(1, 0, 0)),
)
points = np.arange(mesh.npoints)[mesh.x == 1.0]
mpc = fem.MultiPointConstraint(field, points=points, centerpoint=-1, skip=(0, 1, 1))
force = fem.PointLoad(field, points=[-1])

table = fem.math.linsteps([0, 1], num=5, axis=0, axes=3)
ramp = {force: table}
step = fem.Step(items=[solid, mpc, force], ramp=ramp, boundaries=boundaries)

job = fem.Job(steps=[step]).evaluate()
plotter = solid.plot("Principal Values of Cauchy Stress", plotter=mpc.plot())
plotter.show()
```

![image](https://github.com/user-attachments/assets/471baae0-67e0-4c62-9652-6331cec1d682)

If you know your force distribution, you can of course also apply the force on the mesh-points directly.

```python
import felupe as fem
import numpy as np

mesh = fem.Cube(n=6)
region = fem.RegionHexahedron(mesh)
field = fem.FieldContainer([fem.Field(region, dim=3)])
solid = fem.SolidBody(umat=fem.NeoHooke(mu=1, bulk=2), field=field)

boundaries = dict(left=fem.Boundary(field[0], fx=0))
points = np.arange(mesh.npoints)[mesh.x == mesh.x.max()]
force = fem.PointLoad(field, points=points)

table = fem.math.linsteps([0, 1], num=5, axis=0, axes=3)
ramp = {force: 1 / len(force.points) * table}
step = fem.Step(items=[solid, force], ramp=ramp, boundaries=boundaries)

job = fem.Job(steps=[step]).evaluate()
solid.plot("Principal Values of Cauchy Stress").show()
```

![image](https://github.com/user-attachments/assets/5fe10826-90d1-4558-839e-b2916d23d92a)

Hope that helps!

**Reply by KaiWolven** [reply]
Hi,
In your last example, I undersand the applied force is reflected in the "table" variable, in N, if I'm not mistaken, is that correct? Also, in the last image, what units is the "Cauchy Stress" in? In case it is not in the International System of Units.
Thank you!

**Reply by adtzlr** [reply]
Yes, FElupe has no units built in. That means if the

* the model's length unit is mm and 
* the material parameters are in N/mm² (or MPa) then the force unit is N. Cauchy stress is MPa.

Here's a reference: https://www.dynasupport.com/howtos/general/consistent-units

**Reply by adtzlr** [reply]
If you need further assistance, please let me know. Feel free to provide a code snippet with a (partially) working code, I'll try to help you!

**Reply by KaiWolven** [reply]
Hi!

I managed to import a model in VTK (I had to voxelize it in order to make it work, it is a 25mm cube) and apply a load in the negative Z axis (if i'm not mistaken). However, it is not converging properly in the first step and won't continue solving. In case you could offer some insight. The code I'm working on is this:
```
import numpy as np
import pypardiso
import matplotlib.pyplot as plt
import pyvista as pv
import felupe as fem
import meshio

surface = pv.read("Cubo.vtk")
voxels = pv.voxelize(surface, density=surface.length / 20)

mesh = fem.Mesh(
    points=voxels.points,
    cells=voxels.cell_connectivity.reshape(-1, 8),
    cell_type="hexahedron",
)

X, Y, Z = mesh.points.T
mesh.plot().show()

region = fem.RegionHexahedron(mesh)
mesh = mesh.flip(np.any(region.dV < 0, axis=0))

region = fem.RegionHexahedron(mesh)
field = fem.FieldContainer([fem.Field(region, dim=3)])

umat = fem.Hyperelastic(fem.mooney_rivlin, C10=1.594, C01=0.44)
solid = fem.SolidBody(umat, field)

boundaries = dict(left=fem.Boundary(field[0], fz= 0))
points = np.arange(mesh.npoints)[mesh.z == mesh.z.max()]
force = fem.PointLoad(field, points=points)

plotter = pv.Plotter()
actor = plotter.add_points(
    mesh.points[points],
    point_size=20,
    color="red",
)
mesh.plot(plotter=plotter, opacity=0.7).show()

fuerza = -2

table = fem.math.linsteps([0, 1], num=5, axis=2, axes=3)
ramp = {force: fuerza / len(force.points) * table}
step = fem.Step(items=[solid, force], ramp=ramp, boundaries=boundaries)

job = fem.Job(steps=[step]).evaluate()
solid.plot("Principal Values of Cauchy Stress (MPa)").show()
```

**Reply by adtzlr** [reply]
Hi! 

Instead of a voxel mesh, you could also try [`pygmsh`](https://github.com/nschloe/pygmsh), [`scikit-gmsh`](https://github.com/pyvista/scikit-gmsh) or [`tetgen`](https://github.com/pyvista/tetgen) (I have no experience with any of these packages...).

The `Hyperelastic` models like `mooney_rivlin` should be used inside `SolidBodyNearlyIncompressible(umat, bulk=my_bulk_modulus)`. The default `SolidBody` does not add a volumetric strain energy density function. If you prefer a displacement-based formulation for a compressible material, you have to add the volumetric part to the strain energy function.

### a) Compressible materials

```python
import tensortrax.math as tm

def mooney_rivlin(C, C10, C01, bulk):
    U = bulk * tm.log(tm.linalg.det(C))**2 / 8
    return fem.mooney_rivlin(C, C10=C10, C01=C01) + U

umat = fem.Hyperelastic(mooney_rivlin, C10=1.594, C01=0.44, bulk=3.0)
solid = fem.SolidBody(umat, field)
```

### b) Nearly-incompressible materials
> [!NOTE]
> A nearly-incompressible solid body is only available for **Quad** and **Hexahedron** cell types.

```python
umat = fem.Hyperelastic(fem.mooney_rivlin, C10=1.594, C01=0.44)
solid = fem.SolidBodyNearlyIncompressible(umat, field, bulk=5000)
```

To change the name of the Cauchy stress plot, use:
```python
solid.plot(
    name="Principal Values of Cauchy Stress",  # don't use a custom text here
    component=0,  # for other principal components
    label="the first comp. is the max. principal value"  # use this for a custom text
).show()
```

For a voxel-based mesh, you can enhance the matrix-assembly-performance by setting `region = fem.RegionHexahedron(mesh, uniform=True)` (after `mesh.flip(...)` is applied).

**Reply by KaiWolven** [reply]
Hi!

I have to voxelize the mesh in order to make it work as I have not been able to obtain a proper vtk that runs with felupe (the main problem is obtaining an hex mesh I think, instead of tetra). Any recommendations on how to create it?

Thanks!

**Reply by adtzlr** [reply]
Hex-meshing requires either a really good (typically proprietary) software or some manual work. It is not as easy as triangulation. That's one of the reasons why cad-based IGA finite element packages have some popularity.

FElupe supports also tetrahedrons, for nearly-incompressible solids either a) using the MINI-element or b) with mixed P2-P1 quadratic-linear tetrahedrons (similar to [howto/mixed](https://felupe.readthedocs.io/en/latest/howto/mixed.html)).

**Reply by adtzlr** [reply]
Tetra-meshes are of course also supported, even for nearly-incompressible solids.

<details><summary>MINI-element</summary>

```python
import felupe as fem
from functools import partial

mesh = fem.Cube(n=6).triangulate().add_midpoints_volumes(cell_type="tetra")
region = fem.RegionTetraMINI(mesh)
field = fem.FieldsMixed(region, n=3)

boundaries, loadcase = fem.dof.uniaxial(field, clamped=True)

umat = fem.NearlyIncompressible(fem.NeoHooke(mu=1), bulk=5000)
solid = fem.SolidBody(umat, field)

move = fem.math.linsteps([0, 1], num=5)
step = fem.Step(items=[solid], ramp={boundaries["move"]: move}, boundaries=boundaries)

job = fem.CharacteristicCurve(steps=[step], boundary=boundaries["move"])
job.evaluate()

fig, ax = job.plot(
    xlabel="Displacement $d_1$ in mm $\longrightarrow$",
    ylabel="Normal Force $F_1$ in N $\longrightarrow$",
)

solid.plot(
    "Principal Values of Cauchy Stress", project=partial(fem.project, mean=True)
).show()
```

![image](https://github.com/user-attachments/assets/6802777e-1f6b-430a-b4b1-d746781a3ed6)
</details>

<details><summary>P2/P1 tetrahedron</summary>

```python
import felupe as fem
from functools import partial

mesh = fem.Cube(n=6).triangulate().add_midpoints_edges()
region = fem.RegionQuadraticTetra(mesh)
field = fem.FieldsMixed(region, n=3)

boundaries, loadcase = fem.dof.uniaxial(field, clamped=True)

umat = fem.NearlyIncompressible(fem.NeoHooke(mu=1), bulk=5000)
solid = fem.SolidBody(umat, field)

move = fem.math.linsteps([0, 1], num=5)
step = fem.Step(items=[solid], ramp={boundaries["move"]: move}, boundaries=boundaries)

job = fem.CharacteristicCurve(steps=[step], boundary=boundaries["move"])
job.evaluate()

fig, ax = job.plot(
    xlabel="Displacement $d_1$ in mm $\longrightarrow$",
    ylabel="Normal Force $F_1$ in N $\longrightarrow$",
)

solid.plot(
    "Principal Values of Cauchy Stress",
    project=partial(fem.project, mean=True),
    nonlinear_subdivision=3,
    line_width=2,
).show()

```

![image](https://github.com/user-attachments/assets/3ae97ebc-86f6-4b0a-85f2-83d8a125c173)

</detaisl>

**Reply by KaiWolven** [reply]
The main problem with tetra meshes is that I'm not being able to read externally generated ones (I'm using Gmsh to transform already existing stl files), modify the number of elements and operate with it inside Felupe.

**Reply by adtzlr** [reply]
Okay, I understand. There is

```python
import felupe

container = fem.mesh.read("my_mesh.msh")
mesh = container.meshes[0]
```

Does that function work for you?

**Reply by KaiWolven** [reply]
If i start with a closed tetra-mesh like this:

![image](https://github.com/user-attachments/assets/27c2b94c-3919-4113-bff3-c5410dc9efbb)
![image](https://github.com/user-attachments/assets/0a8b5aef-16bc-4844-8b5e-363e6d2a8745)

After applying that function, it seems to read just fine, both in .gms and .vtk:

![image](https://github.com/user-attachments/assets/e90f8140-1b7c-457a-8321-6ff1b0693995)

However, it won't continue working after the fem.RegionTetraMINI(mesh) bit. Printing the mesh class after reading gives "class 'felupe.mesh._mesh.Mesh'", which seems fine. I've tried many approaches when generating the tetra-mesh, but still cannot get it to work properly. I've also tried other Region types in case TetraMINI wasn't compatible. Moreover, if i reduce the opacity when plotting, there seems to be no inner meshing; I donnot know if its just the visualization of there and actual problem when reading the file:

![image](https://github.com/user-attachments/assets/54d2878e-3f3a-4d79-9d5e-879490d9d362)

Hope this explains the problem a little bit better.

**Reply by adtzlr** [reply, marked as answer]
It seems that PyVista doesn't show points inside the volume if you set the opacity to a value lower than 1. I think you'd have to modify the mesh before to do so.

Here is an example. Let's take a gmsh-file, e.g. [cube6.geo](https://www.math.univ-paris13.fr/~cuvelier/software/gmshgeo.html). First, create the msh-file within gmsh and then follow this code to use it in FElupe.

<details><summary>code (requires the pypardiso-package)</summary>

```python
import felupe as fem
import pypardiso

mesh = fem.mesh.read("cube6.msh")[-1]
region = fem.RegionTetraMINI(mesh.add_midpoints_volumes(cell_type="tetra"))
field = fem.FieldsMixed(region, n=3)

boundaries, loadcase = fem.dof.uniaxial(field, clamped=True)

umat = fem.NearlyIncompressible(fem.NeoHooke(mu=1), bulk=5000)
solid = fem.SolidBody(umat, field)

move = fem.math.linsteps([0, 1], num=5)
step = fem.Step(items=[solid], ramp={boundaries["move"]: move}, boundaries=boundaries)

job = fem.CharacteristicCurve(steps=[step], boundary=boundaries["move"])
job.evaluate(solver=pypardiso.spsolve)

fig, ax = job.plot(
    xlabel="Displacement $d_1$ in mm $\longrightarrow$",
    ylabel="Normal Force $F_1$ in N $\longrightarrow$",
)
solid.plot("Principal Values of Cauchy Stress").show()
```

</details>

![image](https://github.com/user-attachments/assets/63cf127a-fa6e-478c-a073-6496b360a721)

**Reply by KaiWolven** [reply]
Thanks for the info! It seems to work just fine with a tetra mesh. The problem was the definition I had for the region part.

I just have one last question: in that example, how would you apply other hyperelastic models appart from NeoHooke, such as Mooney-Rivlin or Ogden? I've tried with fem.Hyperelastic and other previous felupe definitions, but it seems it is not compatible with this workflow.

**Reply by adtzlr** [reply]
That's great @KaiWolven! 🎉 🎉 

To use other hyperelastic materials from FElupe, use

```python
umat = fem.NearlyIncompressible(
    fem.Hyperelastic(fem.mooney_rivlin, C10=0.4, C01=0.1), bulk=5000
)
solid = fem.SolidBody(umat, field)
```

or with a custom strain energy density function:

```python
import tensortrax.math as tm

def my_mooney_rivlin(C, C10, C01):
    I1 = tm.trace(C)
    I2 = (I1**2 - tm.trace(C @ C)) / 2
    I3 = tm.linalg.det(C)
    return C10 * (I3**(-1/3) * I1 - 3) + C01 * (I3**(-2/3) * I2 - 3)
    
umat = fem.NearlyIncompressible(
    fem.Hyperelastic(my_mooney_rivlin, C10=0.4, C01=0.1), bulk=5000
)
solid = fem.SolidBody(umat, field)
```

You have to wrap your model in `felupe.NearlyIncompressible(..., bulk=5000)` to add a volumetric strain energy density function and to create a mixed-field formulation (displacement, pressure, volume-ratio). `SolidBodyNearlyIncompressible(..., bulk=5000)` does the same, but in a more efficient way for quads and hexahedrons.
